tea.py

```diff
@@ -0,0 +1,460 @@
+        if self.parameter_values.get("Positive electrode active material volume fraction") + self.parameter_values.get("Positive electrode porosity") == 1:
+                    self.results["Positive electrode active material mass loading [mg.cm-2]"] = self.results.get("Positive electrode active material mass loading [mg.cm-2]") or self.results.get("Positive electrode active material volume loading [uL.cm-2]") * self.results.get("Positive electrode composite true density [kg.m-3]") / 1000
+        elif (inputs.get("Positive electrode active material true density [kg.m-3]") == None) and (self.results.get("Positive electrode active material true density [kg.m-3]") == None):
+                    self.results["Positive electrode active material mass loading [mg.cm-2]"] = self.results["Positive electrode active material mass loading [mg.cm-2]"] or 0
+                    print("Warning: Missing 'Positive electrode active material true density [kg.m-3]'")
+        else:
+                    self.results["Positive electrode active material mass loading [mg.cm-2]"] = self.results.get("Positive electrode active material mass loading [mg.cm-2]") or self.results.get("Positive electrode active material volume loading [uL.cm-2]") * (inputs.get("Positive electrode active material true density [kg.m-3]") or self.results.get("Positive electrode active material true density [kg.m-3]")) / 1000
+        self.results["Positive electrode electrolyte mass loading [mg.cm-2]"] = self.results.get("Positive electrode electrolyte mass loading [mg.cm-2]") or self.results.get("Positive electrode electrolyte volume loading [uL.cm-2]") * self.results.get("Electrolyte density [kg.m-3]") / 1000
+        self.results["Positive electrode inactive material mass loading [mg.cm-2]"] = self.results.get("Positive electrode inactive material mass loading [mg.cm-2]") or self.results.get("Positive electrode density [kg.m-3]") * self.results.get("Positive electrode thickness [m]") * 100 - self.results.get("Positive electrode active material mass loading [mg.cm-2]")
+        self.results["Positive current collector mass loading [mg.cm-2]"] = self.results.get("Positive current collector mass loading [mg.cm-2]") or self.results.get("Positive current collector volume loading [uL.cm-2]") * self.results.get("Positive current collector density [kg.m-3]") / 1000
+
+        if print_values == True:
+            print(self.results)
+
+    def plot_stack_breakdown(self, inputs=None, print_values=False):
+
+        if inputs is None:
+            inputs = {}
+        else:
+            inputs = inputs
+
+        if print_values == True:
+            print_values = True
+        else:
+            print_values = False
+
+        # call stack_breakdown
```

Comment:
you should track which methods have been called as an attribute of the class so you can skip recalculating everything. at the moment if you call `stack_energy_densities` then `stack_breakdown` then `plot_stack_breakdown` you will have called `stack_energy_densities` 3 times.